air-script/tests/aux_trace/aux_trace.masm

```diff
@@ -0,0 +1,57 @@
+proc.cache_z_exp
+    # Load trace length
+    mem_load.3294967296
+    # => [trace_len, ...]
+    # Push initial num_of_cycles
+    push.1
+    # => [num_of_cycles, trace_len, ...]
+    # Load Z
+    padw mem_loadw.4294967291 drop drop
+    # => [z_1, z_0, num_of_cycles, trace_len, ...]
+    # => [z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+    # Clean stack
+    dropw
+end # END PROC cache_z_exp
+
+proc.cache_periodic_polys
+    exec.cache_z_exp
+end # END PROC cache_periodic_polys
+
+proc.compute_evaluate_transitions
+    padw mem_loadw.4294965000 drop drop padw mem_loadw.4294965001 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.4294965001 movdn.3 movdn.3 drop drop ext2mul padw mem_loadw.4294965002 movdn.3 movdn.3 drop drop ext2mul ext2add ext2sub # constraint 0 for main
+    padw mem_loadw.4294965001 drop drop padw mem_loadw.4294965002 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 drop drop ext2add ext2sub # constraint 1 for main
+    padw mem_loadw.4294965002 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.4294965001 movdn.3 movdn.3 drop drop ext2add ext2sub # constraint 2 for main
+    padw mem_loadw.4294965072 drop drop padw mem_loadw.4294965072 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2add padw mem_loadw.4294965001 movdn.3 movdn.3 drop drop ext2add padw mem_loadw.4294965000 drop drop ext2add ext2mul ext2sub # constraint 0 for aux
+    padw mem_loadw.4294965073 movdn.3 movdn.3 drop drop padw mem_loadw.4294965073 drop drop padw mem_loadw.4294965002 movdn.3 movdn.3 drop drop padw mem_loadw.4294965000 movdn.3 movdn.3 drop drop ext2add ext2mul ext2sub # constraint 1 for aux
+end # END PROC compute_evaluate_transitions
+
+proc.evaluate_transitions
+    exec.cache_periodic_polys
+    exec.compute_evaluate_transitions
```

Comment:
A few nits for the future regarding naming and file organization. I'm thinking that the main exported function which looks something like this:
```
export.evaluate_constraints
  # same as the current cache_periodic_polys
  exec.compute_periodic_values

  # evaluates transition constraints (the result is an aggregated numerator)
  exec.evaluate_transition_constraints

  # evaluates boundary constraints (the result is an aggregated numerator)
  exec.evaluate_boundary_constraints

  # divides numerators by their respective divisors and sums up the evaluations
  exec.combine_constraints
end
```

One other questions: seems like right now we are doing evaluation and aggregation separately. I think this fine for now, but in the future (in a more optimized version), the aggregation result should probably just live on the stack. We may even add a specialized instruction to the VM to make this aggregation simpler.

Reply:
Is there are reason why you want to renamed the procedures?

I used `compute_` for procedures that leave data on the stack, and `cache_` for procedures that save the data to memory.

> One other note: seems like right now we are doing evaluation and aggregation separately. I think this fine for now, but in the future (in a more optimized version), the aggregation result should probably just live on the stack. We may even add a specialized instruction to the VM to make this aggregation simpler.

This was done for testing purposes, otherwise we can't test the intermediate values

Reply:
> Is there are reason why you want to renamed the procedures?
> 
> I used `compute_` for procedures that leave data on the stack, and `cache_` for procedures that save the data to memory.

It communicates the purpose of procedures a bit better I think. Plus, we'll probably have most of these procedures (if not all) write the results to memory rather than leave them on the stack (with `combine_constraints` being potentially the only exception).

> This was done for testing purposes, otherwise we can't test the intermediate values

I think we could still do it by setting random coefficients for specific constraints to zeros. Basically, if we have 3 constraints, and we want to test that the 2nd one evaluates correctly, we could set random coefficient for that constraint to `ONE` and for the other two constraints to `ZERO`. This does assume that there are independent tests which make sure constraint aggregation works correctly.